### DISCUSSION TITLE
Handling route transitions & actions with offline/flaky connections

### DISCUSSION THREAD

**Discussion by azuisleet** [remix-run/remix#6743] [Q&A]
If a user is offline and they attempt change routes, it's very possible that the fetcher will return a `TypeError: Failed to fetch`

Although not addressing this scenario, Remix does attempt to reload the page if the route module isn't found:

https://github.com/remix-run/remix/blob/fbc6e2353b4ef2c2166677b39b7fbd8eb90a5379/packages/remix-react/routeModules.ts#L202

But does _not_ attempt the same if the route module was already cached but the loader failed.

This is similar to other issues like https://github.com/remix-run/remix/discussions/4242, but I believe next.js handles this specific type of navigation better, by reloading the page (so that the UA can show a "No Internet" page), rather than pushing it to an error boundary.

The justification here is that it's fine if useFetcher throws to an error boundary since it's a progressive enhancement, but navigation should be more of a framework concern.

I suppose a very crude workaround would be to handle it post `ErrorBoundary` with something like:

```
  if (error instanceof TypeError && error.message === "Failed to fetch") {
    window.location.reload();
  }
```

**Comment by azuisleet** [comment]
After more consideration, I believe network errors need to be their own class. Here's why:

1. If you explicitly return from a loader/action, and it reaches the user, the application will happily render your page with the loader/action data populated.
2. If you explicitly throw from a loader/action, and it reaches the user, the application will render the nearest page error boundary.
3. If any response fails to reach the user, or they can't reach the service at all, the application will render the nearest page error boundary.

There should at least be some way to handle case 3 gracefully, as say CloudFlare returns a 502 from their edge, and the error boundary can *only* see a generic error with the message `<html> <head><title>502 Bad Gateway</title></head> <body> <center><h1>502 Bad Gateway</h1></center> <hr><center>cloudflare</center> </body> </html>` which isn't all that helpful to display to end users.

It's obviously not possible to gracefully fall back to browser error pages in an SPA, but at a minimum I'd like to be able to:
1. Handle a network related error as a result of a form action, and display a message, regardless of whether the action or subsequent  loaders failed. The use case here is for forms that use controlled inputs that won't have a browser memory, like contenteditable. I suppose you could also argue in favor of storing working input in local/session storage.
2. For non-form submissions, such as transitioning between pages, it would be helpful to know the HTTP status code at the error boundary

For the first case, some users are asking for promise-based handlers for form actions, https://github.com/remix-run/remix/discussions/3955 but it's not clear how that would work if a subsequent loader failed for example. Instead, I'm considering an API that would let routes opt-in to receive the error that would have been thrown to an error boundary via useActionData. For example: https://gist.github.com/azuisleet/f96da706ab741c669fdfc720e8999e3f

For the second, I currently check for TypeErrors at the error boundary and try to render a "No Connection" message. It's not perfect, but I don't have any better ideas.

**Reply by jerrygitsis** [reply]
I found that some browsers also have different values for network errors. I agree it would helpful could if remix could catch these and treat them as a separate class of error in the boundary altogether.
